src/ddmd/dcache.d

```diff
@@ -0,0 +1,377 @@
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/ddmd/dcache.d, _dcache.d)
+ *
+ * Dcache - a inmplementation if a distributed cache over memory-mapped file.
+ * It doesn't try to be smart and has most basic interface in the way of get/put,
+ * both operations are synchronious and "atomic" by using a simple spin-lock.
+ * 
+ * Intended use
+ *
+ * Allow DMD instances (and frontends on its base) to share CTFE computation results and such,
+ * importantly a key must reflect the state of all dependent entities such as contents of files.
+ *
+ * TODOs:
+ * - eviction policy (e.g. time based expiration)
+ * - sizing policy (is 128 Mb enough for everybody?)
+ * - better hash map algorithm (it's really dumb)
+ * - better block management (does a linear pass to clean up entries ATM)
+ */
+module ddmd.dcache;
+
+import core.sys.posix.sys.mman;
+import core.sys.posix.fcntl;
+import core.sys.posix.semaphore;
+import core.sys.posix.time;
+import core.sys.posix.unistd;
+import core.stdc.errno;
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+import core.atomic;
+
+version(Posix)
+{
+        
+    struct DCache
+    {
+        /// Load the cache identified by 'keyspace'
+        void initialize(const(char)[] keyspace)
+        {
+            char[512] buf;
+            const(char)* home = getenv("HOME");
+            if (home == null) home = "/tmp";
+            snprintf(buf.ptr, buf.length-1, "%s/.dcache-%s", home, keyspace.ptr);
+            printf("%s\n", buf.ptr);
+            fd = open(buf.ptr, O_CREAT | O_RDWR, S_IRWXU);
+            if (fd < 0) perror("Failed to open DCache mapping");
+            if (posix_fallocate(fd, 0, DATA_SIZE + TOC_SIZE + META_SIZE) < 0) perror("Failed to fallocate DCache mapping");
+            
+            mapping = mapRange(META_SIZE + DATA_SIZE + TOC_SIZE, fd);
+            owner = cast(shared(pid_t)*)mapping.ptr;
+            toc = cast(TocEntry[])mapping[META_SIZE .. META_SIZE + TOC_SIZE];
+            data = cast(Data*)(mapping.ptr + META_SIZE + TOC_SIZE);
+            if (toc.length & (toc.length-1)) abort();
+            ourPid = getpid();
+            // Lock before modification of shared data
+            {
+                lock();
+                scope(exit) unlock();
+                // If first entry is all zeros - new mapping
+                if (data.occupied == 0 && data.size == 0)
+                {
+                    data.size = DATA_SIZE;
+                }
+            }
+            // Anything older then tsThreshold is considered ripe for eviction
+            tsCompilation = time(null);
+            tsThreshold = tsCompilation - 24*3600;
+        }
+
+        /// Clean the cache
+        void truncate()
+        {
+            lock();
+            scope(exit) unlock();
+            (cast(ubyte[])mapping)[]  = 0;
+            data.size = DATA_SIZE;
+        }
+
+        static struct TocEntry
+        {
+            uint hash32;
```

Comment:
You could store bucket states in the hash.

https://github.com/dlang/druntime/blob/fba1d54da42772bbc128f2a72abca286d2010264/src/rt/aaA.d#L31-L34

Reply:
Good idea.